File: prog-7-2.hs
import Data.Char

endExpr :: String -> Int -> Int
endExpr str ind = go str ind 0
  where
    go [] _ _ = length str
    go (x:xs) i brackets
      | brackets == 0 && (x == ',' || x == ')') = i
      | x == ')' = go xs (i + 1) (brackets - 1)
      | x == '(' = go xs (i + 1) (brackets + 1)
      | otherwise = go xs (i + 1) brackets

logicFunc :: String -> Int -> Bool
logicFunc str ind
  | isLower (str !! ind) = str !! ind == 't'
  | str !! ind == '!' = not (logicFunc str (ind + 2))
  | otherwise = let isAnd = str !! ind == '&'
                    newInd = ind + 2
                in go newInd isAnd
  where
    go i isAnd
      | str !! i == ')' = not isAnd
      | otherwise = let res = logicFunc str (i + 1)
                    in if isAnd && not res
                       then False
                       else if not isAnd && res
                            then True
                            else go (endExpr str (i + 1)) isAnd

main :: IO ()
main = do
  st <- getLine
  print $ logicFunc st 0 

File: prog-7-2.cpp
#include <string>
#include <iostream>
using namespace std;

size_t end_expr (const string& str, size_t ind)
{
	int brackets = 0;

	for (; ind < str.size(); ++ind)
	{
		if (brackets == 0)
		{
			if (str[ind] == ',' or str[ind] == ')')
				return ind;
		}

		if (str[ind] == ')')
			--brackets;

		if (str[ind] == '(')
			++brackets;	
	}

	return str.size();
}

bool logic_func (const string& str, size_t ind)
{
	if (str[ind] == 't' or str[ind] == 'f')
		return str[ind] == 't';

	if (str[ind] == '!')
		return not logic_func(str, ind + 2);

	bool is_AND = str[ind] == '&';
	ind += 2;

	while (str[ind] != ')')
	{
		bool res = logic_func(str, ind + 1);

		if (is_AND and not res)
		{
			return false;
		}

		if (not is_AND and res)
		{
			return true;
		}

		ind = end_expr(str, ind + 1);
	}

	return not is_AND;
}

int main()
{
	string st;
	cin >> st;
	cout << logic_func(st, 0) << endl;
}